common/protob/messages-cardano.proto

```diff
@@ -109,13 +118,224 @@ message CardanoPublicKey {
+    optional uint64 ttl = 8;                                    // transaction ttl - added in shelley
+    required uint32 network_id = 9;                            // network id - mainnet or testnet
+    optional uint64 validity_interval_start = 10;
+    required bool has_auxiliary_data = 11;
+    required uint32 witnesses_count = 12;
+}
+
+/**
+ * Request: Transaction input data
+ * @next CardanoTxItemAck
+ */
+message CardanoTxInput {
+    repeated uint32 address_n = 1;  // BIP-32 path to derive the key from master node
+    required bytes prev_hash = 2;   // hash of previous transaction output to spend by this input
+    required uint32 prev_index = 3; // index of previous output to spend
+}
+
+/**
+ * Request: Transaction output data
+ * @next CardanoTxItemAck
+ */
+message CardanoTxOutput {
+    optional string address = 1;                                    // target coin address in bech32 or base58
+    required uint64 amount = 2;                                     // amount to spend
+    optional CardanoAddressParametersType address_parameters = 3;   // parameters used to derive the address
+    required uint32 asset_groups_count = 4;
+}
+
+/**
+ * Request: Transaction output asset group data
+ * @next CardanoTxItemAck
+ */
+message CardanoAssetGroup {
+    required bytes policy_id = 1;                  // asset group policy id
+    required uint32 tokens_count = 2;
+}
+
+/**
+ * Request: Transaction output asset group token data
+ * @next CardanoTxItemAck
+ */
+message CardanoToken {
+    required bytes asset_name_bytes = 1;     // asset name as bytestring (may be either ascii string or hash)
+    required uint64 amount = 2;              // asset amount
+}
+
+/**
+ * Request: Stake pool owner parameters
+ * @next CardanoTxItemAck
+ */
+message CardanoPoolOwner {
+    repeated uint32 staking_key_path = 1;   // BIP-32-style path to derive staking key of the owner
+    optional bytes staking_key_hash = 2;    // owner's staking key if it is an external owner
+}
+
+/**
+ * Request: Stake pool relay parameters
+ * @next CardanoTxItemAck
+ */
+message CardanoPoolRelayParameters {
+    required CardanoPoolRelayType type = 1;   // pool relay type
+    optional bytes ipv4_address = 2;          // ipv4 address of the relay given as 4 bytes
+    optional bytes ipv6_address = 3;          // ipv6 address of the relay given as 16 bytes
+    optional string host_name = 4;            // relay host name given as URL, at most 64 characters
+    optional uint32 port = 5;                 // relay port number in the range 0-65535
+}
+
+/**
+ * Stake pool metadata parameters
+ */
+message CardanoPoolMetadataType {
+    required string url = 1;   // stake pool url hosting metadata, at most 64 characters
+    required bytes hash = 2;   // stake pool metadata hash
+}
+
+/**
+ * Stake pool parameters
+ */
+message CardanoPoolParametersType {
+    required bytes pool_id = 1;                                          // stake pool cold public key hash (28 bytes)
```

Comment:
We use `required` and `optional` to mean what they actually mean and they result in different python types. With `optional`, `| None` is added to the type. Since AFAIK `required` can easily be changed to `optional` without breaking anything, I'd leave this as `required` and it can be made `optional` when we decide to add signing by pool operators.